pkg/server/plugin/upstreamauthority/gcpca/gcpca.go

```diff
@@ -0,0 +1,332 @@
+	if err := hcl.Decode(config, req.Configuration); err != nil {
+		return nil, pluginErr.New("unable to decode configuration: %v", err)
+	}
+
+	// Swap out the current configuration with the new configuration
+	p.setConfig(config)
+
+	return &plugin.ConfigureResponse{}, nil
+}
+
+func (p *Plugin) GetPluginInfo(ctx context.Context, req *plugin.GetPluginInfoRequest) (*plugin.GetPluginInfoResponse, error) {
+	return &plugin.GetPluginInfoResponse{}, nil
+}
+
+func (p *Plugin) getConfig() (*Config, error) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	if p.c == nil {
+		return nil, pluginErr.New("not configured")
+	}
+
+	return p.c, nil
+}
+
+func (p *Plugin) setConfig(c *Config) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	p.c = c
+}
+
+func getClient(ctx context.Context) (*pcaapi.CertificateAuthorityClient, error) {
+	creds, err := google.FindDefaultCredentials(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	pcaClient, err := pcaapi.NewCertificateAuthorityClient(ctx, option.WithCredentials(creds))
+	if err != nil {
+		return nil, err
+	}
+
+	return pcaClient, nil
+}
+
+func pemToX509(pemStr string) (*x509.Certificate, error) {
+	block, rest := pem.Decode([]byte(pemStr))
+	if len(rest) > 0 {
+		return nil, errors.New("Unexpected trailer in certificate of size " + fmt.Sprint(len(rest)))
+	}
+
+	cert, err := x509.ParseCertificate(block.Bytes)
+	if err != nil {
+		return nil, err
+	}
+
+	return cert, nil
+}
+
+func loadCertificateAuthorities(ctx context.Context, client *pcaapi.CertificateAuthorityClient, spec CertificateAuthoritySpec) ([]*privatecapb.CertificateAuthority, error) {
+	allCerts := make([]*privatecapb.CertificateAuthority, 0)
+	certIt := client.ListCertificateAuthorities(ctx, &privatecapb.ListCertificateAuthoritiesRequest{
+		Parent: spec.caParentPath(),
+		Filter: "labels." + spec.LabelKey + ":" + spec.LabelValue,
+	})
+
+	p := iterator.NewPager(certIt, 20, "")
+	for {
+		var page []*privatecapb.CertificateAuthority
+
+		nextPageToken, err := p.NextPage(&page)
+		if err != nil {
+			return nil, err
+		}
+
+		allCerts = append(allCerts, page...)
+		if nextPageToken == "" {
+			break
+		}
+	}
+
+	return allCerts, nil
+}
+
+func (p *Plugin) mintX509CA(ctx context.Context, csr []byte, preferredTtl int32) (*upstreamauthority.MintX509CAResponse, error) {
+	csrParsed, err := x509.ParseCertificateRequest(csr)
+	if err != nil {
+		return nil, fmt.Errorf("unable to parse CSR: %v", err)
+	}
+
+	validity := time.Second * time.Duration(preferredTtl)
+
+	pcaClient, err := getClient(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	allCertRoots, err := loadCertificateAuthorities(ctx, pcaClient, p.c.RootSpec)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to load root CAs: %v.", err)
+	}
+	if len(allCertRoots) == 0 {
+		return nil, fmt.Errorf("No certificate authorities found with label pair %s:%s.", p.c.RootSpec.LabelKey, p.c.RootSpec.LabelValue)
+	}
+
+	// The certs get returned in issuance order, so just grab the last one and use that.
```

Comment:
I was thinking something more like the following:
- We always have 2 roots stamped with the appropriate label
- We specify some period p
- Every p, some process creates a new root as a duplicate of the latest root in the pair
- The process then disables the oldest root in the pair

Spire has a default TTL of its intermediate certs, and auto-rotates them every TTL/2.  That autorotation will requery CAS, find the latest of the two roots, and issue a cert chaining to that root.  The other root goes in the trust bundle.

No restarts needed!

Reply:
If we're duplicating the latest root, is this only to handle the case of an expiring root? What about a compromised key, when immediate rotation is necessary?

During the autorotation, the latest of the two roots has not yet been distributed in a trust bundle to all agents. Should we instead be grabbing the first/oldest of the two roots, so that the latest root gets distributed to all agents before use? 

* Process `p` creates a new root
* `p` disables the oldest root
* On auto rotation, SPIRE starts issuing off the now-oldest root and includes the new root in the bundle.

This also handles the case of compromise - If the oldest is comprised, then the new root has already been distributed. If the new root is compromised, that's also fine, since we're issuing off the oldest root still. (If both are comprised, obviously this would require a full restart)

Reply:
Using the oldest also gives you the benefit of being able to test out the latest root, without it being in use yet.

Reply:
That makes sense to me.  Oldest root it is then.  Thank you!